### PR TITLE
feat: Add keeper methods for historacle prices and medians

### PR DESCRIPTION
**Pull request by rbajollari**
## Description

closes: #1540

---

### Author Checklist

_All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues._

I have...

- [ ] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added `!` to the type prefix if API or client breaking change
- [ ] added appropriate labels to the PR
- [ ] targeted the correct branch (see [PR Targeting](https://github.com/umee-network/umee/blob/main/CONTRIBUTING.md#pr-targeting))
- [ ] provided a link to the relevant issue or specification
- [ ] added a changelog entry to `CHANGELOG.md`
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

_All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items._

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed all author checklist items have been addressed
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

File: x/oracle/keeper/historic_price.go
Comment (adamewozniak) [inline x/oracle/keeper/historic_price.go:23]: So we sort by rates, but what if the rates are the same? Is it randomized order?

File: x/oracle/keeper/historic_price.go
Comment (adamewozniak) [inline x/oracle/keeper/historic_price.go:26]:
```suggestion
	if lenPrices % 2 == 0 {
```

File: x/oracle/keeper/historic_price.go
Comment (adamewozniak) [inline x/oracle/keeper/historic_price.go:165]: You'll want to pass in the exchangeRate rather than querying for it again in `setMedian`
```suggestion
	k.setMedian(ctx, denom, exchangeRate)
```

File: x/oracle/keeper/historic_price.go
Comment (adamewozniak) [inline x/oracle/keeper/historic_price.go:97]: This will be consumed by ABCI endblocker where they're already uppercase, or through the querier, where we can just set it to uppercase. No need to put it in this keeper imo 
```suggestion
```

File: x/oracle/keeper/historic_price.go
Comment (adamewozniak) [inline x/oracle/keeper/historic_price.go:129]: You might want to put this denom parser in a util function

File: x/oracle/keeper/historic_price.go
Comment (adamewozniak) [inline x/oracle/keeper/historic_price.go:136]: This is a bit repetitive, in another PR do you think it would make sense to make the `HistoricPrice` struct look like this?

```golang
type HistoricPrice struct {
	Rate ExchangeRateTuple
	BlockNum      uint64
}
```

File: x/oracle/keeper/historic_price.go
Comment (adamewozniak) [inline x/oracle/keeper/historic_price.go:182]: Keeper comment please 😄 

Also, we'll have to prune historic prices after a certain amount of time rather than just clearing all of them out, so idk if we need this. We can add that keeper function in another PR

File: x/oracle/keeper/params.go
Comment (adamewozniak) [inline x/oracle/keeper/params.go:66]:
```suggestion
// StampPeriod returns the amount of blocks the historacle endblocker waits
```

File: x/oracle/keeper/params.go
Comment (adamewozniak) [inline x/oracle/keeper/params.go:73]:
```suggestion
// SetStampPeriod updates the amount of blocks the historacle endblocker waits
```

Same here, technically historacle isn't a module of its own ^

File: x/oracle/types/params.go
Comment (adamewozniak) [inline x/oracle/types/params.go:169]: perfect 💯

File: x/oracle/types/params.go
Comment (adamewozniak) [inline x/oracle/types/params.go:330]:
```suggestion
	if v < 1 {
		return fmt.Errorf("stamp period must be positive: %d", v)
	}

	return nil
}

func validatePrunePeriod(i interface{}) error {
	v, ok := i.(uint64)
	if !ok {
		return fmt.Errorf("invalid parameter type: %T", i)
	}

	if v < 1 {
		return fmt.Errorf("prune period must be positive: %d", v)
	}
```

**Review by adamewozniak — COMMENTED** [review]
initial review - great job @rbajollari !!

File: x/oracle/keeper/historic_price.go
Comment (robert-zaremba) [inline x/oracle/keeper/historic_price.go:34]:
```suggestion
	if lenPrices%2 == 0 {
		return prices[lenPrices/2-1].ExchangeRates.ExchangeRate.
			Add(prices[lenPrices/2].ExchangeRates.ExchangeRate).
			QuoInt64(2)
	}
	return prices[lenPrices/2].ExchangeRates.ExchangeRate
```

File: x/oracle/keeper/historic_price.go
Comment (robert-zaremba) [inline x/oracle/keeper/historic_price.go:66]:
```suggestion
	store := ctx.KVStore(k.storeKey)
	historicPrices := k.GetHistoricPrices(ctx, denom)
	median := median(historicPrices)
```

File: x/oracle/keeper/historic_price.go
Comment (robert-zaremba) [inline x/oracle/keeper/historic_price.go:103]: Why don't we create a list of exchange rates only? We don't need to repeat `denom` in all elements.

File: x/oracle/keeper/historic_price.go
Comment (robert-zaremba) [inline x/oracle/keeper/historic_price.go:96]: pagination must be added, otherwise it will blow up the tx time.

File: x/oracle/keeper/historic_price.go
Comment (robert-zaremba) [inline x/oracle/keeper/historic_price.go:134]:
```suggestion
		var historicPrice types.HistoricPrice
```

File: x/oracle/keeper/historic_price.go
Comment (robert-zaremba) [inline x/oracle/keeper/historic_price.go:116]:
```suggestion
// denom in the store.
// Iterator stops when exhausting the source, or when the handler returns `true`.
```

File: x/oracle/keeper/historic_price.go
Comment (robert-zaremba) [inline x/oracle/keeper/historic_price.go:152]: why storing `denom`? It's already in the key. Let's avoid bloating the state.

File: x/oracle/types/keys.go
Comment (robert-zaremba) [inline x/oracle/types/keys.go:64]:
```suggestion
func MedianKey(denom string) (key []byte) {
```

Comment (robert-zaremba) [reply x/oracle/types/keys.go:64]: We are not getting anything from anywhere.

File: x/oracle/types/keys.go
Comment (robert-zaremba) [inline x/oracle/types/keys.go:71]:
```suggestion
func HistoricPriceKey(denom string, blockNum uint64) (key []byte) {
```

File: x/oracle/keeper/historic_price.go
Comment (robert-zaremba) [inline x/oracle/keeper/historic_price.go:192]: We should have an option to efficiently delete prices older then some time (eg older than 2 weeks ago). Otherwise this will cause significant storage use.

File: x/oracle/keeper/params.go
Comment (robert-zaremba) [inline x/oracle/keeper/params.go:66]:
```suggestion
// StampPeriod returns the amount of blocks the oracle module waits
```

File: x/oracle/keeper/params.go
Comment (robert-zaremba) [inline x/oracle/keeper/params.go:73]:
```suggestion
// SetStampPeriod updates the amount of blocks the oracle module waits
```

File: x/oracle/types/params.go
Comment (robert-zaremba) [inline x/oracle/types/params.go:31]: What if a block time will change? It can be updated without changing a code, through a gov voting.

File: x/oracle/keeper/historic_price.go
Comment (robert-zaremba) [inline x/oracle/keeper/historic_price.go:96]: pagination must be added, otherwise it will blow up the tx time.

File: x/oracle/keeper/historic_price.go
Comment (robert-zaremba) [inline x/oracle/keeper/historic_price.go:165]: this is super inefficient. You read all historical prices and recalculate median during every price update.

**Review by robert-zaremba — CHANGES_REQUESTED** [review]
Blocking because:
* adding historical price is super inefficient (see comment)
* have concerns about state bloat: how the state will be managed and cleaned?
* why do we have one big median? Shouldn't we have median per "epoch"?

Is there any spec clarifying the pruning functionality? If not, I can help with that if needed.

File: x/oracle/keeper/historic_price.go
Comment (robert-zaremba) [reply x/oracle/keeper/historic_price.go:23]: Median is median. It's OK to have same values in a sequence.

File: x/oracle/keeper/historic_price.go
Comment (rbajollari) [reply x/oracle/keeper/historic_price.go:135]: I think it already is like that. Do you mean like this?
```golang
type HistoricPrice struct {
	Rate          sdk.Dec
	BlockNum      uint64
}
```

File: x/oracle/keeper/historic_price.go
Comment (rbajollari) [reply x/oracle/keeper/historic_price.go]: I put ParseDemonFromHistoricPriceKey and ParseBlockFromHistoricPriceKey as their own methods in keys.go rather than in the utils folder since these don't seem general enough to belong in there. Lmk what you think about that.

File: x/oracle/keeper/historic_price.go
Comment (rbajollari) [reply x/oracle/keeper/historic_price.go]: Removed for now and just going to keep the DeleteHistoricPrice method for this PR

File: x/oracle/keeper/historic_price.go
Comment (adamewozniak) [reply x/oracle/keeper/historic_price.go:126]: Yeah, if the denom is a part of the key we don't need to read this

File: x/oracle/keeper/historic_price.go
Comment (rbajollari) [reply x/oracle/keeper/historic_price.go]: This is now changed to not get called inside the keeper and be used like outlined in the spec. SetMedian will get called every MedianPeriod rather than every time a new price is added.

File: x/oracle/keeper/historic_price.go
Comment (rbajollari) [reply x/oracle/keeper/historic_price.go]: Proto changed to be just exchangerate rather than tuple 👍

File: x/oracle/types/keys.go
Comment (rbajollari) [reply x/oracle/types/keys.go]: While I don't disagree with the naming update the rest of the methods in this file have "Get" in front of them, so updating just these two methods would look inconsistent. I'd suggest doing this in another PR.

File: x/oracle/types/params.go
Comment (adamewozniak) [reply x/oracle/types/params.go:32]: Then we'll have to change these definitions for the entire oracle module, that's not a part of this PR

File: x/oracle/keeper/historic_price.go
Comment (adamewozniak) [reply x/oracle/keeper/historic_price.go:139]: this `GetHistoricPrices` method will likely only be called internally. @rbajollari wdyt?

Comment (rbajollari) [reply x/oracle/keeper/historic_price.go:139]: Updated to only iterate price keys with the given denom 👍

File: x/oracle/types/keys.go
Comment (adamewozniak) [reply x/oracle/types/keys.go]: I agree @rbajollari

File: proto/umee/oracle/v1/oracle.proto
Comment (adamewozniak) [inline proto/umee/oracle/v1/oracle.proto:115]: We don't need yaml tags anymore:
```suggestion
```

File: proto/umee/oracle/v1/oracle.proto
Comment (adamewozniak) [inline proto/umee/oracle/v1/oracle.proto:50]:
```suggestion
  // Median Period represents the amount blocks we will wait between
```

File: x/oracle/keeper/historic_price.go
Comment (adamewozniak) [inline x/oracle/keeper/historic_price.go:58]: It might make more sense to do a new error message here, since initially medians will not exist until the first `Median Period` passes

File: x/oracle/keeper/historic_price.go
Comment (adamewozniak) [inline x/oracle/keeper/historic_price.go:93]: Same here, might want another error message

File: x/oracle/keeper/historic_price.go
Comment (adamewozniak) [inline x/oracle/keeper/historic_price.go:118]: Does this need to be a public keeper function?
```suggestion
func (k Keeper) getHistoricPrice(
```

File: x/oracle/keeper/historic_price.go
Comment (adamewozniak) [inline x/oracle/keeper/historic_price.go:136]: Same
```suggestion
func (k Keeper) getHistoricPrices(
```

File: x/oracle/keeper/historic_price.go
Comment (adamewozniak) [inline x/oracle/keeper/historic_price.go:178]:
```suggestion
// AddHistoricPrice adds the historic price of a denom at the current
// block height.
```

File: x/oracle/keeper/historic_price.go
Comment (adamewozniak) [inline x/oracle/keeper/historic_price.go:204]:
```suggestion
// DeleteMedian deletes a given denom's median price in the last prune
```

**Review by adamewozniak — APPROVED** [review]
this looks good to me, just a few comments. 

I'd like to have a review from @zarazan , and maybe @toteki or @RafilxTenfen as well 🙏

File: x/oracle/keeper/historic_price.go
Comment (rbajollari) [reply x/oracle/keeper/historic_price.go]: This method isn't called within the keeper so since it doesn't need to be public I've removed it

**Comment by adamewozniak** [comment]
waiting for a review from @RafilxTenfen to merge 🙏

File: x/oracle/types/keys.go
Comment (RafilxTenfen) [inline x/oracle/types/keys.go:71]: We do this in the same way 3 times, maybe add a function to it?

**Review by RafilxTenfen — APPROVED** [review]
Just one comment
utACK :100:

File: x/oracle/keeper/historic_price.go
Comment (adamewozniak) [reply x/oracle/keeper/historic_price.go:121]: Will be addressed in #1570

**Pull request merged**
